File: packages/dotnet/src/MyGrainExchange.Api/Overlay/MgxApiError.cs
/* Hand-written overlay — not generated. */
using System;
using System.Collections.Generic;
using System.Text.Json;
using MyGrainExchange.Api.Client;

namespace MyGrainExchange.Api.Overlay
{
    /// <summary>A single field-level validation error.</summary>
    public sealed class MgxFieldError
    {
        public string Field { get; }
        public string Message { get; }

        public MgxFieldError(string field, string message)
        {
            Field = field;
            Message = message;
        }
    }

    /// <summary>
    /// Typed error mapped from the API's <c>{ "error": { status, code, message, errors[] } }</c>
    /// envelope. Thrown by every <see cref="MgxClient"/> method instead of the generated
    /// <see cref="ApiException"/>.
    /// </summary>
    public sealed class MgxApiError : Exception
    {
        /// <summary>HTTP status code (or the status echoed in the error envelope).</summary>
        public int Status { get; }

        /// <summary>Machine-readable error code (e.g. <c>validation_failed</c>).</summary>
        public string Code { get; }

        /// <summary>Field-level validation errors, when present.</summary>
        public IReadOnlyList<MgxFieldError> FieldErrors { get; }

        public MgxApiError(int status, string code, string message, IReadOnlyList<MgxFieldError> fieldErrors = null)
            : base(message)
        {
            Status = status;
            Code = code;
            FieldErrors = fieldErrors ?? Array.Empty<MgxFieldError>();
        }

        /// <summary>
        /// Builds an <see cref="MgxApiError"/> from the generated <see cref="ApiException"/>,
        /// parsing its <see cref="ApiException.ErrorContent"/> (the raw response body) as the
        /// MGX error envelope. Falls back to the HTTP status when the body is missing or not JSON.
        /// </summary>
        public static MgxApiError FromApiException(ApiException ex)
        {
            int status = ex.ErrorCode;
            string code = "error";
            string message = string.IsNullOrEmpty(ex.Message) ? "Request failed" : ex.Message;
            var fieldErrors = new List<MgxFieldError>();

            var body = ex.ErrorContent as string ?? ex.ErrorContent?.ToString();
            if (!string.IsNullOrWhiteSpace(body))
            {
                try
                {
                    using var doc = JsonDocument.Parse(body);
                    if (doc.RootElement.ValueKind == JsonValueKind.Object &&
                        doc.RootElement.TryGetProperty("error", out var error) &&
                        error.ValueKind == JsonValueKind.Object)
                    {
                        if (error.TryGetProperty("status", out var s) && s.ValueKind == JsonValueKind.Number)
                            status = s.GetInt32();
                        if (error.TryGetProperty("code", out var c) && c.ValueKind == JsonValueKind.String)
                            code = c.GetString();
                        if (error.TryGetProperty("message", out var m) && m.ValueKind == JsonValueKind.String)
                            message = m.GetString();
                        if (error.TryGetProperty("errors", out var errs) && errs.ValueKind == JsonValueKind.Array)
                        {
                            foreach (var item in errs.EnumerateArray())
                            {
                                if (item.ValueKind != JsonValueKind.Object) continue;
                                var field = item.TryGetProperty("field", out var f) && f.ValueKind == JsonValueKind.String
                                    ? f.GetString() : string.Empty;
                                var fm = item.TryGetProperty("message", out var im) && im.ValueKind == JsonValueKind.String
                                    ? im.GetString() : string.Empty;
                                fieldErrors.Add(new MgxFieldError(field, fm));
                            }
                        }
                    }
                }
                catch (JsonException)
                {
                    /* non-JSON error body — keep the HTTP status/message. */
                }
            }

            return new MgxApiError(status, code, message, fieldErrors);
        }
    }
}

File: packages/php/overlay/MgxApiError.php
<?php

/* Hand-written overlay — not generated. */

declare(strict_types=1);

namespace MyGrainExchange\Mgx\Overlay;

use MyGrainExchange\Mgx\ApiException;

/**
 * Typed error mapped from the API's
 * `{ "error": { status, code, message, errors[] } }` envelope.
 *
 * Field-level validation errors are exposed as a list of
 * `['field' => string, 'message' => string]` arrays via {@see fieldErrors()}.
 */
class MgxApiError extends \Exception
{
    /** @var int HTTP status code. */
    private int $status;

    /** @var string Machine-readable error code (e.g. "validation_error"). */
    private string $errorCode;

    /** @var list<array{field: string, message: string}> */
    private array $fieldErrors;

    /**
     * @param list<array{field: string, message: string}> $fieldErrors
     */
    public function __construct(
        int $status,
        string $code,
        string $message,
        array $fieldErrors = [],
        ?\Throwable $previous = null,
    ) {
        parent::__construct($message, $status, $previous);
        $this->status = $status;
        $this->errorCode = $code;
        $this->fieldErrors = $fieldErrors;
    }

    /** HTTP status code of the failed response. */
    public function status(): int
    {
        return $this->status;
    }

    /** Machine-readable error code from the API envelope. */
    public function code(): string
    {
        return $this->errorCode;
    }

    /** @return list<array{field: string, message: string}> */
    public function fieldErrors(): array
    {
        return $this->fieldErrors;
    }

    /**
     * Build an {@see MgxApiError} from a generated {@see ApiException}, parsing the
     * `{ error: { status, code, message, errors[] } }` envelope when present.
     */
    public static function fromApiException(ApiException $e): self
    {
        $status = (int) $e->getCode();
        if ($status === 0) {
            $status = 500;
        }
        $code = 'error';
        $message = $e->getMessage() !== '' ? $e->getMessage() : 'Request failed';
        $fieldErrors = [];

        $body = self::decodeBody($e->getResponseBody());
        if (is_array($body) && isset($body['error']) && is_array($body['error'])) {
            $error = $body['error'];
            if (isset($error['status'])) {
                $status = (int) $error['status'];
            }
            if (isset($error['code']) && is_string($error['code'])) {
                $code = $error['code'];
            }
            if (isset($error['message']) && is_string($error['message'])) {
                $message = $error['message'];
            }
            if (isset($error['errors']) && is_array($error['errors'])) {
                foreach ($error['errors'] as $fieldError) {
                    if (!is_array($fieldError)) {
                        continue;
                    }
                    $fieldErrors[] = [
                        'field' => (string) ($fieldError['field'] ?? ''),
                        'message' => (string) ($fieldError['message'] ?? ''),
                    ];
                }
            }
        }

        return new self($status, $code, $message, $fieldErrors, $e);
    }

    /**
     * Normalize the generated exception's response body (which may be a JSON
     * string, an \stdClass, or already an array) into an associative array.
     *
     * @param mixed $body
     * @return array<string, mixed>|null
     */
    private static function decodeBody($body): ?array
    {
        if (is_string($body)) {
            $decoded = json_decode($body, true);

            return is_array($decoded) ? $decoded : null;
        }
        if ($body instanceof \stdClass) {
            $decoded = json_decode((string) json_encode($body), true);

            return is_array($decoded) ? $decoded : null;
        }
        if (is_array($body)) {
            return $body;
        }

        return null;
    }
}
